all-packages: move aliases to aliases.nix

matthewbauer opened the pull request:
Lots of aliases have been left over. This is a massive edit that
changes many random lines.

Another spinoff from #39515.

matthewbauer committed:
all-packages: move aliases to aliases.nix

Lots of aliases have been left over. This is a massive edit that
changes many random lines.

matthewbauer committed:
all-packages: normalize overrides

Overrides should only be on the "default" package not the other way around.

matthewbauer committed:
aliases: add gccApple

matthewbauer committed:
aliases: add check if alias still exists in all-packages.nix

If an alias is already defined in all-packages.nix, we want to throw
an error to make it obvious that something is wrong. Otherwise there
is no way to realize that the alias is shadowing the real definition.

matthewbauer committed:
aliases: add skip aliases config

You can turn on this config option if you want to find references to
aliases in Nixpkgs. Ideally these can be removed from Nixpkgs and
eventually we can remove the alias altogether.

Ericson2314 reviewed (approved):
Looks great!

matthewbauer commented on pkgs/top-level/stage.nix:
```diff
@@ -97,7 +97,8 @@ let
     in res;
 
-  aliases = self: super: import ./aliases.nix super;
+  aliases = self: super: if config.skipAliases or false then {}
```
Yes, we could even put a warning in aliases as well. It would probably be too annoying right now though- aliases are still used extensively in Nixpkgs.

oxij commented:
Awesome.

I put the following into `check-sameness.nix`

```nix
with import ./nixpkgs/lib/default.nix;

let
  pkgs = import ./nixpkgs/default.nix {};

  pkgsA = import ./nixpkgs_old/default.nix {};
  pkgsB = import ./nixpkgs_new/default.nix {};
  a = pkgs.writeText "pkgsA-list" (concatStringsSep "\n" (attrNames pkgsA));
  b = pkgs.writeText "pkgsB-list" (concatStringsSep "\n" (attrNames pkgsB));
in
  pkgs.runCommand "pkgs-diff" {
    inherit a b;
  } ''
    diff -Nua "$a" "$b" > $out || true
  ''
```

checked out before and after trees into `nixpkgs_old` and `nixpkgs_new`,
built the derivation, and got this:

```screen
--- /nix/store/<censored>-pkgsA-list	1970-01-01 00:00:01.000000000 +0000
+++ /nix/store/<censored>-pkgsB-list	1970-01-01 00:00:01.000000000 +0000
@@ -216,6 +216,7 @@
 ansible_2_5
 ansifilter
 ant
+ant_1_9
 antfs-cli
 anthy
 antigen
```

Is this a bug?

globin commented:
@oxij from the diff that looks like it's on purpose to align it with the unversioned `ant` attribute.

@matthewbauer feel free to hit merge after resolving conflicts!

matthewbauer commented on pkgs/top-level/aliases.nix:
```diff
@@ -17,17 +17,31 @@ let
 
+  # Make sure that we are not shadowing something from
+  # all-packages.nix.
+  checkInPkgs = n: alias: if builtins.hasAttr n self
```
aliases is called with "import ./aliases.nix super" - so I think it works okay. Not sure if this is incorrect though.

matthewbauer committed:
mysql → mariadb

matthewbauer commented:
In case anyone is interested, here is the script I'm using to find dupes in all-packaages.nix:

```
grep 'callPackage ' pkgs/top-level/all-packages.nix | sed 's,^.*callPackage ,,' | sed 's, .*,,' | sort | uniq -d
```

matthewbauer committed:
all-packages: more overrides normalizing

matthewbauer committed:
all-packages: cleanup from last commit

matthewbauer commented:
@GrahamcOfBorg eval

matthewbauer committed:
all-packages: revert 2 name changes

I am reverting two name changes that might not be a good idea:

- gnum4 → m4
- apacheAnt → ant

These are debatable changes & not sure what’s best. I prefered the
short version because there are not alternatives- but will not hold
off for now.

volth commented:
Would be nice to clean-up aliased (and thus deprecating) names treewide to reduce confusions
```s/kerberos/libkrb5/g``` etc

volth commented:
What I do not like in this PR:

There was a difference between "aliases" in ```all-packages.nix``` and in ```aliases.nix```
The former are mostly "moving parts" which can be overwritten in overlays (```python2 = python27``` as an example) while the latter are for phasing out names (```new-name = deprecatedName```) and thus immutable.

I am not sure that for example ```mysql = mariadb``` (and other names from this PR) is not a moving part

matthewbauer commented:
@volth Yes I can definitely see the case w.r.t. mysql. Feel free to move any back that don't make sense. It's hard to make that judgement on so many aliases. My basic rule was to leave any "versioned" attributes but that was the only exception.

volth commented:
A thumb rule could be buildability of ```nixpkgs``` with ```config.skipAliases = true```.

The changes required to build ```nixpkgs``` with ```config.skipAliases = true``` would reveal should ```mariadb``` leak to places where generic ```mysql``` is enough, etc (much like "should python27 leak where python2 is enough")

This would untangle weird cases like using ```kerberos``` (it is often a derivation's argument name (for example ```curl```), sometimes overwritten to ```libkrb5``` in "all-packages.nix", sometimes defaulting to ```libkrb5``` via "aiases.nix"; I am not sure how it should be fixed properly

matthewbauer commented:
> A thumb rule could be buildability of nixpkgs with config.skipAliases = true.

I would agree at least if you substitute "buildability" with "eval". Aliases should not affect the building of anything.

Right now I think skipAliases will always fail but we can definitely improve tht.

volth commented:
>> A thumb rule could be buildability of nixpkgs with config.skipAliases = true.

> I would agree at least if you substitute "buildability" with "eval". Aliases should not affect the building of anything.

Yes

> Right now I think skipAliases will always fail but we can definitely improve that.

Actually, I had something like ```skipAliases``` in my nixpkgs after issues with overriding XFCE aliases (https://github.com/NixOS/nixpkgs/pull/33600 https://github.com/NixOS/nixpkgs/pull/36888) and made PRs which maintained that <s>build</s>```eval```-ability (https://github.com/NixOS/nixpkgs/pull/37616 https://github.com/NixOS/nixpkgs/pull/36891 https://github.com/NixOS/nixpkgs/pull/36889)

This PR introduced changes which are not so easy to fix.
There are only 18 of them which break ```eval```:
```
kerberos = libkrb5;
procps = procps-ng;
yacc = bison;
devicemapper = lvm2;
libudev = udev;
udev = systemd;
lzma = xz;
dbus_libs = dbus;
docbook5_xsl = docbook_xsl_ns;
libsysfs = sysfsutils;
nologin = shadow;
fam = gamin;
telnet = inetutils;
mysql = mariadb;
openal = openalSoft;
pulseaudioLight = pulseaudio;
x11 = xlibsWrapper;
gmock = gtest;
```

and if ```mysql``` and ```telnet``` could be moved back to "all-packages.nix" and ```openal```, ```pulseaudioLight```, ```x11```, ```gmock``` ... can be replaced with new names treewide, I am not sure what to do with ```udev```, ```kerberos```, ```procps```, ```lzma```
